tests/cluster-check.sh

```diff
@@ -0,0 +1,118 @@
+#!/bin/bash
+
+# in case the scheduler goes nuts, wrap ourselves in a timeout so
+# there's a bounded completion time
+if test "$CHECK_CLUSTER_SUBPROCESS" = ""; then
+   export CHECK_CLUSTER_SUBPROCESS=1
+   timeout -s KILL 30m /bin/bash ./cluster-check.sh "$@"
+   exit $?
+fi
+
+scheduler="$1"
+
+echo "--> scheduler: $scheduler"
+
+set -e
+
+# we submit 2 2-node jobs, each of which are a sleep for 16 minutes.
+# The whole thing has to run in 30 minutes, or the kill above will
+# fail the job, which means that the jobs must run at the same time.
+# The initial cluster is 2 nodes, so we'll need to scale up 2 nodes in
+# less than 10 minutes in order for the test to succeed.
+
+if test "$scheduler" = "slurm" ; then
+    cat > job1.sh <<EOF
+#!/bin/bash
+srun sleep 960
+touch job1.done
+EOF
+    cat > job2.sh <<EOF
+#!/bin/bash
+srun sleep 960
+touch job2.done
+EOF
+
+    chmod +x job1.sh job2.sh
+    rm -f job1.done job2.done
+
+    sbatch -N 2 ./job1.sh
+    sbatch -N 2 ./job2.sh
+
+    done=0
+    while test $done = 0 ; do
+	if test -f job1.done -a -f job2.done; then
+	    done=1
+	else
+	    sleep 1
+	fi
+    done
+
+elif test "$scheduler" = "sge" ; then
+    ppn=` qhost | grep 'lx-' | head -n 1 | sed -n -e 's/[^[:space:]]*[[:space:]]\+[^[:space:]]*[[:space:]]\+\([0-9]\+\).*/\1/p'`
+    count=$((ppn * 2))
+
+    cat > job1.sh <<EOF
+#!/bin/bash
+#$ -pe mpi $count
+#$ -R y
+
+sleep 960
+touch job1.done
+EOF
+    cat > job2.sh <<EOF
+#!/bin/bash
+#$ -pe mpi $count
+#$ -R y
+
+sleep 960
+touch job2.done
+EOF
+
+    chmod +x job1.sh job2.sh
+    rm -f job1.done job2.done
+
+    qsub ./job1.sh
+    qsub ./job2.sh
+
+    done=0
+    while test $done = 0 ; do
+        if test -f job1.done -a -f job2.done; then
+            done=1
+        else
+            sleep 1
+        fi
+    done
+
+elif test "$scheduler" = "torque" ; then
+    cat > job1.sh <<EOF
+#!/bin/bash
+sleep 1200
+touch job1.done
+EOF
+    cat > job2.sh <<EOF
+#!/bin/bash
+sleep 1200
+touch job2.done
+EOF
+
+    chmod +x job1.sh job2.sh
+    rm -f job1.done job2.done
+
+    qsub -l nodes=2:ppn=2 ./job1.sh
+    qsub -l nodes=2:ppn=2 ./job2.sh
+
+    done=0
+    while test $done = 0 ; do
```

Comment:
While loop in each schedulers can moved out of if/else block. can avoid duplicate code here.